{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.1 CellChat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(CellChat)\n",
    "library(patchwork)\n",
    "library(dplyr)\n",
    "library(Seurat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Cellchat calculations are performed with deconvoluted data\n",
    "## Key code are showen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assignLabels <- function(object, prediction = \"predictions\") {\n",
    "  pred <- object[[prediction]]@data\n",
    "  pred <- pred[1:(nrow(pred)-1), ]\n",
    "  # label each spot based on the maximum prediction probability\n",
    "  labels = rownames(pred)[apply(pred, 2, which.max)]\n",
    "  names(labels) <- colnames(pred)\n",
    "  object$labels <- factor(labels)\n",
    "  Idents(object) <- \"labels\"\n",
    "  return(object)\n",
    "}\n",
    "\n",
    "Sol <- assignLabels(spe.Sol2, prediction = \"predictions\")\n",
    "EDL <- assignLabels(spe.EDL2, prediction = \"predictions\")\n",
    "LD <- assignLabels(spe.LD2, prediction = \"predictions\")\n",
    "\n",
    "data.input1 = Seurat::GetAssayData(Sol, slot = \"data\", assay = \"Spatial\") \n",
    "data.input2 = Seurat::GetAssayData(EDL, slot = \"data\", assay = \"Spatial\") \n",
    "data.input3 = Seurat::GetAssayData(LD, slot = \"data\", assay = \"Spatial\") \n",
    "\n",
    "genes1 <- rownames(data.input1)\n",
    "genes2 <- rownames(data.input2)\n",
    "genes3 <- rownames(data.input3)\n",
    "\n",
    "genes.common <- Reduce(intersect, list(genes1, genes2, genes3))\n",
    "\n",
    "colnames(data.input1) <- paste0(\"A1_\", colnames(data.input1))\n",
    "colnames(data.input2) <- paste0(\"A2_\", colnames(data.input2))\n",
    "colnames(data.input3) <- paste0(\"A3_\", colnames(data.input3))\n",
    "data.input <- cbind(data.input1[genes.common, ], data.input2[genes.common, ], data.input3[genes.common, ])\n",
    "\n",
    "# define the meta data\n",
    "meta1 = data.frame(labels = Idents(Sol), samples = \"A1\") # manually create a dataframe consisting of the cell labels\n",
    "meta2 = data.frame(labels = Idents(EDL), samples = \"A2\") \n",
    "meta3 = data.frame(labels = Idents(LD), samples = \"A3\") \n",
    "\n",
    "meta <- rbind(meta1, meta2, meta3)\n",
    "rownames(meta) <- colnames(data.input)\n",
    "\n",
    "meta$labels <- factor(meta$labels, levels = levels(Idents(Sol)))\n",
    "meta$samples <- factor(meta$samples, levels = c(\"A1\", \"A2\", \"A3\"))\n",
    "unique(meta$labels) \n",
    "unique(meta$samples)\n",
    "\n",
    "# load spatial transcriptomics information\n",
    "Sol_positions <- read_csv(\"Sol_positions.csv\")\n",
    "LD_positions <- read_csv(\"LD_positions.csv\")\n",
    "EDL_positions <- read_csv(\"EDL_positions.csv\")\n",
    "\n",
    "Sol_positions <- as.data.frame(Sol_positions)\n",
    "LD_positions <- as.data.frame(LD_positions)\n",
    "EDL_positions <- as.data.frame(EDL_positions)\n",
    "\n",
    "rownames(Sol_positions) <- Sol_positions[[1]]\n",
    "Sol_positions <- Sol_positions[ , -1]\n",
    "\n",
    "rownames(LD_positions) <- LD_positions[[1]]\n",
    "LD_positions <- LD_positions[ , -1]\n",
    "\n",
    "rownames(EDL_positions) <- EDL_positions[[1]]\n",
    "EDL_positions <- EDL_positions[ , -1]\n",
    "\n",
    "spatial.locs_Sol <- Sol_positions[, c(\"pxl_row_in_fullres\", \"pxl_col_in_fullres\")]\n",
    "spatial.locs_EDL <- EDL_positions[, c(\"pxl_row_in_fullres\", \"pxl_col_in_fullres\")]\n",
    "spatial.locs_LD <- LD_positions[, c(\"pxl_row_in_fullres\", \"pxl_col_in_fullres\")]\n",
    "\n",
    "rownames(spatial.locs_Sol) <- paste0(\"A1_\", rownames(spatial.locs_Sol))\n",
    "rownames(spatial.locs_EDL) <- paste0(\"A2_\", rownames(spatial.locs_EDL))\n",
    "rownames(spatial.locs_LD) <- paste0(\"A3_\", rownames(spatial.locs_LD))\n",
    "\n",
    "spatial.locs <- rbind(spatial.locs_Sol, spatial.locs_EDL, spatial.locs_LD)\n",
    "rownames(spatial.locs) <- colnames(data.input)\n",
    "\n",
    "matching_rows <- rownames(spatial.locs) %in% colnames(data.input)\n",
    "cat(\"Number of matching rows: \", sum(matching_rows), \"\\n\")\n",
    "spatial.locs <- spatial.locs[matching_rows, ]\n",
    "rownames(spatial.locs) <- colnames(data.input)[match(rownames(spatial.locs), colnames(data.input))]\n",
    "\n",
    "scalefactors1 = jsonlite::fromJSON(txt = file.path(\"Z:/A/ST/Analysis/Results7/cellchatv2\", 'Sol_scalefactors_json.json'))\n",
    "scalefactors2 = jsonlite::fromJSON(txt = file.path(\"Z:/A/ST/Analysis/Results7/cellchatv2\", 'EDL_scalefactors_json.json'))\n",
    "scalefactors3 = jsonlite::fromJSON(txt = file.path(\"Z:/A/ST/Analysis/Results7/cellchatv2\", 'LD_scalefactors_json.json'))\n",
    "spot.size = 65\n",
    "\n",
    "conversion.factor1 = spot.size/scalefactors1$spot_diameter_fullres\n",
    "conversion.factor2 = spot.size/scalefactors2$spot_diameter_fullres\n",
    "conversion.factor3 = spot.size/scalefactors3$spot_diameter_fullres\n",
    "\n",
    "spatial.factors1 = data.frame(ratio = conversion.factor1, tol = spot.size/2)\n",
    "spatial.factors2 = data.frame(ratio = conversion.factor2, tol = spot.size/2)\n",
    "spatial.factors3 = data.frame(ratio = conversion.factor3, tol = spot.size/2)\n",
    "\n",
    "spatial.factors <- rbind(spatial.factors1, spatial.factors2, spatial.factors3)\n",
    "rownames(spatial.factors) <- c(\"A1\", \"A2\", \"A3\")\n",
    "\n",
    "cellchat <- createCellChat(object = data.input, meta = meta, group.by = \"labels\", datatype = \"spatial\", coordinates = spatial.locs, spatial.factors = spatial.factors)\n",
    "## ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.2 Gene expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Fiugure 6H\n",
    "plot_gene_expression <- function(seurat_object, genes, colors, border_size = 0.1, border_color = \"black\", \n",
    "                                 min_alpha = 0.1, r = 100) {\n",
    "  expression_data <- FetchData(seurat_object, vars = genes)\n",
    "  coords <- GetTissueCoordinates(seurat_object, image = \"slice1\")\n",
    "  plot_data <- cbind(coords, expression_data)\n",
    "  plot_data_long <- plot_data %>%\n",
    "    pivot_longer(\n",
    "      cols = all_of(genes),\n",
    "      names_to = \"gene\",\n",
    "      values_to = \"expression\"\n",
    "    )\n",
    "  p <- ggplot(plot_data_long) +\n",
    "    geom_arc_bar(\n",
    "      aes(\n",
    "        x0 = x, y0 = y,\n",
    "        r0 = 0, r = r, \n",
    "        start = case_when(\n",
    "          gene == genes[1] ~ 0,\n",
    "          gene == genes[2] ~ 2 * pi / 3,\n",
    "          gene == genes[3] ~ 4 * pi / 3\n",
    "        ),\n",
    "        end = case_when(\n",
    "          gene == genes[1] ~ 2 * pi / 3,\n",
    "          gene == genes[2] ~ 4 * pi / 3,\n",
    "          gene == genes[3] ~ 2 * pi\n",
    "        ),\n",
    "        fill = gene,  \n",
    "        alpha = pmax(expression, min_alpha)  \n",
    "      ),\n",
    "      color = border_color, \n",
    "      size = border_size  \n",
    "    ) +\n",
    "    scale_fill_manual(\n",
    "      values = colors \n",
    "    ) +\n",
    "    scale_alpha_continuous(range = c(min_alpha, 1)) +  \n",
    "    coord_fixed() +\n",
    "    scale_y_reverse() + \n",
    "    theme_void() +\n",
    "    theme(legend.position = \"right\") +\n",
    "    labs(fill = \"Gene\", alpha = \"Expression Level\")\n",
    "  print(p)\n",
    "}\n",
    "\n",
    "plot_gene_expression(\n",
    "  seurat_object = EDL,\n",
    "  genes = c(\"BMP7\", \"BMPR1A\", \"BMPR2\"),\n",
    "  colors = c(\"BMP7\" = \"red\", \"BMPR1A\" = \"green\", \"BMPR2\" = \"blue\"),\n",
    "  border_size = 0.2,\n",
    "  border_color = \"black\", \n",
    "  min_alpha = 0.01,\n",
    "  r = 150\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
