{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 Upstream analysis by cellranger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "export PATH=/storage/public/home/2008116199/yxd/soft/cellranger/cellranger-7.1.0:$PATH\n",
    "cellranger testrun --id=tiny\n",
    "/storage/public/home/2008116199/yxd/soft/cellranger/cellranger-8.0.1/bin/cellranger testrun --id=tiny\n",
    "mkdir ~/yard/run_cellranger_mkfastq\n",
    "cellranger mkgtf Sus_scrofa.Sscrofa11.1.109.gtf Sus_scrofa.Sscrofa11.1.109.filtered.gtf --attribute=key:allowable_value\n",
    "cellranger mkref --genome=Sus_scrofa --fasta=/home/yixudong/REF/Sus_scrofa.Sscrofa11.1.dna.toplevel.fa --genes=/home/yixudong/REF/cellranger/Sus_scrofa.Sscrofa11.1.109.chr.filtered.gtf\n",
    "cellranger mkref --genome=Danio.rerio_genome --fasta=Danio_rerio.GRCz11.dna.primary_assembly.fa \\\n",
    "  --genes=Danio_rerio.GRCz11.105.filtered.gtf\n",
    "cellranger count --id=run_count_LD --fastqs=/home/yixudong/sn/LD --sample=LD-1,LD-2,LD-3 --transcriptome=/home/yixudong/sn/SOL/Sus_scrofa\n",
    "/home/yixudong/software/cellranger/cellranger-7.1.0/bin/cellranger count --id=run_count_LD --fastqs=/home/yixudong/Data/muscle --sample=WHB5CDNAPEP00008519 --transcriptome=/home/yixudong/sn/SOL/Sus_scrofa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 QC and Cell annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(SeuratData)\n",
    "library(patchwork)\n",
    "library(limma)\n",
    "library(dplyr)\n",
    "library(magrittr)\n",
    "library(clustree)\n",
    "\n",
    "IntegrateAndProcessData <- function(\n",
    "  data_dirs = c(\n",
    "    Sol = \"Z:\\\\A\\\\ST\\\\Analysis\\\\SnRNAseq\\\\run_count_SOL\\\\outs\\\\filtered_feature_bc_matrix\",\n",
    "    EDL = \"Z:\\\\A\\\\ST\\\\Analysis\\\\SnRNAseq\\\\run_count_EDL\\\\outs\\\\filtered_feature_bc_matrix\",\n",
    "    LD  = \"Z:\\\\A\\\\ST\\\\Analysis\\\\SnRNAseq\\\\run_count_LD\\\\outs\\\\filtered_feature_bc_matrix\"\n",
    "  )\n",
    ") {\n",
    "  library(Seurat)\n",
    "  process_single_dataset <- function(data_dir, project_name) {\n",
    "    counts <- Read10X(data.dir = data_dir)\n",
    "    obj <- CreateSeuratObject(counts = counts, project = project_name, min.cells = 3, min.features = 200)\n",
    "    obj[[\"percent.mt\"]] <- PercentageFeatureSet(obj, pattern = \"^MT-\")\n",
    "    obj <- NormalizeData(obj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
    "    obj <- FindVariableFeatures(obj, selection.method = \"vst\", nfeatures = 2000)\n",
    "    return(obj)\n",
    "  }\n",
    "  datasets <- lapply(names(data_dirs), function(name) {\n",
    "    process_single_dataset(data_dir = data_dirs[[name]], project_name = name)\n",
    "  })\n",
    "  names(datasets) <- names(data_dirs)\n",
    "  for(name in names(datasets)) {\n",
    "    VlnPlot(datasets[[name]], features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), pt.size=0, cols=\"#CCC9E6\", ncol = 3)  \n",
    "  }\n",
    "  my36colors <- c('#CE0317', '#009959', '#3AD7CC', '#993591', '#D6E7A3', '#57C3F3', '#476D87',\n",
    "                  '#E95C59', '#E59CC4', '#AB3282', '#23452F', '#BD956A', '#8C549C', '#585658',\n",
    "                  '#9FA3A8', '#E0D4CA', '#5F3D69', '#C5DEBA', '#58A4C3', '#E4C755', '#F7F398',\n",
    "                  '#AA9A59', '#E63863', '#E39A35', '#C1E6F3', '#6778AE', '#91D0BE', '#B53E2B',\n",
    "                  '#712820', '#DCC1DD', '#CCE0F5', '#CCC9E6', '#625D9E', '#68A180', '#3A6963',\n",
    "                  '#968175')\n",
    "  plot_list <- list()\n",
    "  for(name in names(datasets)) {\n",
    "    top20 <- head(VariableFeatures(datasets[[name]]), 20)\n",
    "    p <- VariableFeaturePlot(datasets[[name]], pt.size = 3)\n",
    "    p <- LabelPoints(plot = p, points = top20, repel = TRUE)\n",
    "    plot_list[[name]] <- p\n",
    "  }\n",
    "  return(list(\n",
    "    datasets = datasets,\n",
    "    plots = plot_list,\n",
    "    colors = my36colors\n",
    "  ))\n",
    "}\n",
    "\n",
    "# result <- IntegrateAndProcessData()\n",
    "\n",
    "\n",
    "MuScle.list.Sc=list(Sol_Sc,EDL_Sc,LD_Sc)\n",
    "\n",
    "MuScle.list.Sc <- lapply(X = MuScle.list.Sc, FUN = function(x) {\n",
    "    x <- NormalizeData(x)\n",
    "    x <- FindVariableFeatures(x, selection.method = \"vst\", nfeatures = 2000)\n",
    "})\n",
    "\n",
    "features.Sc <- SelectIntegrationFeatures(object.list = MuScle.list.Sc)\n",
    "MuScle.Sc.anchors <- FindIntegrationAnchors(object.list = MuScle.list.Sc, anchor.features = features.Sc)\n",
    "MuScle.list.Sc <- IntegrateData(anchorset = MuScle.Sc.anchors)\n",
    "\n",
    "DefaultAssay(MuScle.list.Sc) <- \"integrated\"\n",
    "FindClusters(MuScle.list.Sc, resolution = seq(0.1, 0.5, 0.05))\n",
    "\n",
    "DimPlot(MuScle.list.Sc, reduction = \"umap\", group.by = \"orig.ident\",cols=c(\"#CE0317\",\"#EFD06A\",\"#009959\"), pt.size=0.5)\n",
    "DimPlot(MuScle.list.Sc, reduction = \"umap\", label = TRUE, repel = TRUE, cols=my36colors, pt.size=0.5)\n",
    "DimPlot(MuScle.list.Sc, reduction = \"umap\", label = TRUE, repel = TRUE, cols=my36colors, pt.size=0.5, split.by=\"orig.ident\")\n",
    "\n",
    "head(AverageExpression(object = MuScle.list.Sc))\n",
    "AverageExp <- AverageExpression(object = MuScle.list.Sc)\n",
    "typeof(AverageExp)\n",
    "head(AverageExp$RNA)\n",
    "custom_row_order <- c(\"0\", \"3\", \"1\", \"9\",\"12\", \"7\", \"19\", \"20\", \"2\", \"4\", \"14\",\"16\", \"17\", \"13\", \"15\", \"8\", \"10\", \"5\", \"11\", \"6\", \"18\")\n",
    "custom_col_order <- c(\"0\", \"3\", \"1\", \"9\",\"12\", \"7\", \"19\", \"20\", \"2\", \"4\", \"14\",\"16\", \"17\", \"13\", \"15\", \"8\", \"10\", \"5\", \"11\", \"6\", \"18\")\n",
    "ordered_matrix <- coorda$r[custom_row_order, custom_col_order]\n",
    "coorda<-corr.test(AverageExp$RNA,AverageExp$RNA,method = \"spearman\")\n",
    "colors <- colorRampPalette(c(\"white\",\"red\"))(100)\n",
    "pheatmap(ordered_matrix,color = colors,cluster_rows = FALSE, cluster_cols = FALSE)\n",
    "pheatmap(coorda$r,color = colors)\n",
    "\n",
    "total.prop=prop.table(table(Idents(MuScle.list.Sc)))\n",
    "total.prop <- as.data.frame(total.prop)\n",
    "ggplot(total.prop, aes(x = \"\", y = Freq, fill = Var1)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  theme_minimal() +\n",
    "  scale_fill_manual(values = my36colors)+\n",
    "  labs(x = NULL, y = \"Ratio\", fill = \"Sample Type\", title = \"Sample Ratio by Type\") +\n",
    "  theme(axis.title.x = element_blank(),\n",
    "        axis.text.x = element_blank(),\n",
    "        axis.ticks.x = element_blank())\n",
    "table(Idents(MuScle.list.Sc), MuScle.list.Sc$orig.ident)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 SigleR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_matrix <- GetAssayData(SOL5M, assay = \"RNA\", slot = \"counts\")\n",
    "sce_ref <- SingleCellExperiment(assays = list(counts = expr_matrix))\n",
    "logcounts <- log1p(expr_matrix)\n",
    "assays(sce_ref)$logcounts <- logcounts\n",
    "labels <- as.character(SOL5M@active.ident)\n",
    "\n",
    "MuScle.list.SOL5M <- SingleR(test = MuScle.listSingleR, ref = sce_ref, labels = labels)\n",
    "\n",
    "table(MuScle.list.SOL5M$labels, MuScle.list$seurat_clusters)\n",
    "\n",
    "MuScle.list@meta.data$labels <- MuScle.list.SOL5M$labels\n",
    "DimPlot(MuScle.list, group.by = c(\"seurat_clusters\", \"labels\"), reduction = \"pca\")\n",
    "\n",
    "process_seurat_object <- function(seurat_obj) {\n",
    "  library(Seurat)\n",
    "  library(SingleCellExperiment)\n",
    "  expr_matrix <- GetAssayData(seurat_obj, assay = \"RNA\", slot = \"counts\")\n",
    "  expr_matrix@Dimnames[[1]] <- toupper(expr_matrix@Dimnames[[1]])\n",
    "  sce_ref <- SingleCellExperiment(assays = list(counts = expr_matrix))\n",
    "  logcounts_data <- log1p(expr_matrix)\n",
    "  assays(sce_ref)$logcounts <- logcounts_data\n",
    "  labels <- as.character(seurat_obj@active.ident)\n",
    "  return(list(sce = sce_ref, labels = labels))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
